### PR TITLE
8244531: Tests: add support to identify recurring issues with controls et al

### PR DESCRIPTION
**Pull request by kleopatra**
It's a task to support cross-control/skin testing for recurring issues (like memory leaks on switching skins)

Basically, there's a utility class
- to access lists of all control/classes,
- to access/create all behaviors
- has alternative skins classes for all controls and utility to install/replace skins

POC:
- added tests Skin/BehaviorMemoryTest as examples for using
- changed SkinDisposeContractTest to use

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8244531](https://bugs.openjdk.java.net/browse/JDK-8244531): Tests: add support to identify recurring issues with controls et al

### Reviewers
 * Ajit Ghaisas ([aghaisas](@aghaisas) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/223/head:pull/223`
`$ git checkout pull/223`

### CODE REVIEW

**Comment by aghaisas** [comment]
I have not reviewed this PR yet, but a quick comment - Can we enable  and fix currently Ignored ControlSkinTest with this?

**Comment by kleopatra** [comment]
>  Can we enable and fix currently Ignored ControlSkinTest with this?

will check, always ignored that test :)

**Comment by kleopatra** [comment]
direct re-enable probably is not possible, but I think we can re-write most of it (not sure about the css stuff) - that would be a follow-up issue.

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Example of writing a test for a streak of similar issues,
+ * here f.i. memory leak in behavior.
+ *
+ * Note: doesn't have postfix _Test_ to not be included in
+ * default testing.
+ */
+@RunWith(Parameterized.class)
+public class ControlBehaviorTestExample {
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java:52]: This test does not execute when run as part of :controls:test. We need to rename this class (and file) name to end with 'Test'. I tried doing it and lot of failures are seen.

Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java:52]: that's the point - they are examples of how-to use it and see our tasks :)

Will address your other comments in a minute - and add two real tests (on skin/behavior memory leaks with the misbehaving classes removed: then we'll have both the passing classes and informal ignores on what's going wrong.

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            assertSame(controlClass, controls.get(i));
+        }
+    }
+
+    @Test
+    public void testAlternativeSkinAssignable() {
+        for (int i = 0; i < controlClasses.length; i++) {
+            Class<Control> controlClass = (Class<Control>) controlClasses[i][0];
+            Control control = createControl(controlClass);
+            Skin<?> old = replaceSkin(control);
```
[the remark this answers is not in the document]
Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactoryTest.java:91]: oops ... forgot, will do

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactory.java [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+        return ControlSkinShim.getBehavior(skin);
+    }
+
+    /**
+     * Creates and returns the default behavior for the given control.
+     *
+     * @param <T> the type of the control
+     * @param control the control to create the behavior for
+     * @return the default behavior for the control
+     * @throws RuntimeException with the exception thrown when instantiating the behavior
```
[the remark this answers is not in the document]
Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactory.java:212]: hmm .. the runtime is wrapping the checked exceptions, now tried to clarify by "RuntimeException wrapped around"

File: modules/javafx.controls/src/shims/java/javafx/scene/control/skin/ControlSkinShim.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+package javafx.scene.control.skin;
+
+import java.lang.reflect.Field;
+
+import com.sun.javafx.scene.control.behavior.BehaviorBase;
+
+import javafx.scene.control.Skin;
+
+/**
+ *
```
[the remark this answers is not in the document]
Comment (kleopatra) [reply modules/javafx.controls/src/shims/java/javafx/scene/control/skin/ControlSkinShim.java]: removed

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactoryTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlSkinFactoryTest.java:91]: done

File: modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java (hunk 1 above)
Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java:52]: added SkinMemoryLeakTest/BehaviorMemoryLeakTest as (more or less ;) outlined in the bug report: in the parameter list we start with all controls/classes and then remove the failing explicitly. Future fixes of the leaking classes will have to remove the fixed from the excluded. 

Hmm ..  now probably should remove the xxTestExamples (there existence doesn't seem to have been clear enough anyway ;)

Comment (aghaisas) [reply modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/infrastructure/ControlBehaviorTestExample.java:52]: Yes. It is better to remove this xxTestExamples file.

**Comment by kleopatra** [comment]
done - both removed the example and updated the PR comment

**Comment by aghaisas** [comment]
PR looks fine to me.
Thanks for coming up with this general test for Skins and Behaviors. Good contribution!!

**Pull request closed without merging**